RestartRequested Update API Review (#4733)
MicrosoftEdge/WebView2Feedback: specs/RestartRequested.md

Restart Requested
===

# Background

> There are often times when WebView2 need to be restarted to apply certain update, the purpose of 

Reviewer: Running through copilot - fixing small grammar issues

```suggestion
There are often times when WebView2 needs to be restarted to apply certain updates or change configuration. The purpose of this API is to detect whether such a restart is requested and to provide the urgency of the restart. WebView2 developers can listen to this event for version updates, version downgrades, or important configuration changes to determine if they need to prompt the user for a restart to apply those updates.
```

Restart Requested
===

# Background
There are often times when WebView2 needs to be restarted to apply certain updates or change 
configuration. The purpose of this API is to detect whether such a restart is requested and 
to provide the urgency of the restart. WebView2 developers can listen to this event for version 
updates, version downgrades, or important configuration changes to determine if they need to 
prompt the user for a restart to apply those updates.

> This API is to detect whether a restart is requested from WebView2 base on different priority levels.

Author: Removed

        webView.CoreWebView2.Environment.RestartRequested += WebView_RestartRequested;
    }
}

void WebView_RestartRequested(CoreWebView2Environment sender, object e)
{
    // Depending on your app experience, you can let user
    // to restart on normal cadence.
    RemainderToRestartForUpdate();
}
```

### Scenario 2: High Priority
```c#
void WebView_CoreWebView2InitializationCompleted(object sender, CoreWebView2InitializationCompletedEventArgs e)
{
    if (e.IsSuccess)
    {
        // ...
        webView.CoreWebView2.Environment.RestartPriorityLevel = CoreWebView2RestartPriorityLevel.Critical;
        webView.CoreWebView2.Environment.RestartRequested += WebView_RestartRequested;
    }
}

void WebView_RestartRequested(CoreWebView2Environment sender, object e)
{
    // With ciritical priority, you should prompt user to save their current state and restart the app.
    RestartIfSelectedByUser();
}
```

## Win32 C++
### Scenario 1: Normal Priority
```cpp
wil::com_ptr<ICoreWebView2Environment> m_webViewEnvironment;
void CoreWebView2InitializationCompleted() {
    auto env10 = m_webViewEnvironment.try_query<ICoreWebView2Environment10>();
    if (env10) {

>         CHECK_FAILURE(env10->put_RestartPriorityLevel(COREWEBVIEW2_RESTART_PRIORITY_LEVEL_BEST_EFFORT));

Reviewer: Scenario sample 1 & 2 should be combined into one event handler that checks the Priority in the event handler to decide if the update should be applied immediately or not.

    }
}

void WebView_RestartRequested(CoreWebView2Environment sender, CoreWebView2RestartRequestedEventArgs e)
{
    if (e.Priority == RestartRequestedPriority.BestEffort) 
    {

>         // Depending on your app experience, you should remind user

Reviewer: ```suggestion
        // Depending on your app experience, you should remind the user
```

>         // Depending on your app experience, you should remaind user

Reviewer 2: remaind => remind (there's a few occurrences of this)

        // to restart on normal cadence.

>         RemainderToRestartForUpdate();

Reviewer: ```suggestion
        RemindToRestartForUpdate();
```

    }
    else
    {
        // Depending on your app experience, you should prompt

>         // user to save their current state and restart.

Reviewer: ```suggestion
        // the user to save their current state and restart or otherwise more urgently restart the WebView2.
```

    }
}

void WebView_RestartRequested(CoreWebView2Environment sender, CoreWebView2RestartRequestedEventArgs e)
{

>     if (e.Priority == RestartRequestedPriority.Normal) 

Reviewer 3: ```suggestion
    if (e.Priority == CoreWebView2RestartRequestedPriority.Normal) 
```

Reviewer: Please fix.

>     if (e.Priority == RestartRequestedPriority.Normal) 

Reviewer 4: Not sure how much you want to future-proof. This sample is assuming it will stay a two state enum. You could make this a <= check (maybe change Normal to be above zero), and leave room to have multiple forms of high pri.

Alternatively, if it's really going to stay a two-state, then make it a bool to avoid confusion.

Reviewer 5: Windows OS and WASDK have examples where Lower-than-normal priority enums have negative values, but I don't know if that's appropriate for WV2.

(e.g. https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.ui.dispatching.dispatcherqueuepriority?view=windows-app-sdk-1.5 )

Using an "IsHighPriotity" bool now still allows us to add a 3+ level enum property later if needed, but we'd have to partition them into "High=true" and "High=false" buckets at that time.

Reviewer 5: I think either bool or <= check is OK.  I'm slightly inclined toward <=, and use a positive number for .Normal if we don't support negative values.

I'm inclined to define an enum now instead of using a bool, because I can already think of two possible values we'd add later:  "Low" and a higher-than-"High" value if we wanted to be able to recommend restarting without even waiting for the user prompt mentioned in this sample.

Reviewer 6: Or we can leave gaps in the enum so we can slot new values later.

```
typedef enum COREWEBVIEW2_RESTART_REQUESTED_PRIORITY {
  /// Developer should remind user to restart.
  COREWEBVIEW2_RESTART_REQUESTED_PRIORITY_NORMAL = 1000,
  /// Developer should prompt user to restart as soon as possible. 
  COREWEBVIEW2_RESTART_REQUESTED_PRIORITY_HIGH = 2000,
} COREWEBVIEW2_RESTART_REQUESTED_PRIORITY;
```

Then we can later add `_LOW` as 500 or `_CRITICAL` as 3000.

Maintaining the numeric sequencing allows the app to use greater-than/less-than testing.

Reviewer 4: Frank's [DispatcherQueuePriority](https://docs.microsoft.com/uwp/api/Windows.System.DispatcherQueuePriority) example actually uses both negative numbers and VB spacing :).

Both sound good to me. I wasn't sure if negative numbers would work, because enums can technically be signed or unsigned, but we have a working example. Maybe in the WinRT projection it's always signed.

Reviewer 6: Each language decides whether enums are signed. In C# they are always signed. In JavaScript they are always signed. In C++ they can be either (C++/WinRT projects enums as signed, except that flags enums project as unsigned). I wouldn't be surprised if there's some language that doesn't support negative enums. (It looks like Rust is one of those, but rust-rt works around it by just projecting enums as integer constants!)

Even if signed enums are allowed (or worked around) in all languages, I wouldn't put it past people to get the cast wrong and accidentally cast to unsigned instead of signed.

Reviewer 4: I forgot about [flags], they're unsigned in the c# projection too.

Reviewer: Sounds good. Talking with folks there is a possibility of future enum values. We can go with the suggested values and have plenty of room in the future and switch the sample code to use `<= Normal`

```
typedef enum COREWEBVIEW2_RESTART_REQUESTED_PRIORITY {
  /// Developer should remind user to restart.
  COREWEBVIEW2_RESTART_REQUESTED_PRIORITY_NORMAL = 1000,
  /// Developer should prompt user to restart as soon as possible. 
  COREWEBVIEW2_RESTART_REQUESTED_PRIORITY_HIGH = 2000,
} COREWEBVIEW2_RESTART_REQUESTED_PRIORITY;
```

Author: Updated.

    {
        // Depending on your app experience, you should remind the user
        // to restart on normal cadence.
        RemindToRestartForUpdate();
    }

>     else

Reviewer 5: Don't assume "!=Normal" means "High".  This encourages code for which if we defined a "Low" Priority enum value later, would treat High and Low the same.  (presumably the developer would want Low to behave like "Normal" in this sample case.)

Reviewer: This is discussed in the above comment, thx

Author: Updated

interface ICoreWebView2Environment10;
interface ICoreWebView2RestartRequestedEventHandler;

/// The restart prioirty level can be set by the `RestartPriorityLevel` property.
/// The default value is `COREWEBVIEW2_RESTART_PRIORITY_LEVEL_BEST_EFFORT`.
[v1_enum]
typedef enum COREWEBVIEW2_RESTART_PRIORITY_LEVEL {
  /// Dveloper should prompt user to restart on best effort.
  COREWEBVIEW2_RESTART_PRIORITY_LEVEL_BEST_EFFORT,
  /// Developer should prompt user to restart as soon as possible. 

>   c,

Reviewer: Typo

[uuid(ef7632ec-d86e-46dd-9d59-e6ffb5c87878), object, pointer_default(unique)]
interface ICoreWebView2Environment10 : IUnknown {
  /// Add an event handler for the `RestartRequested` event.
  /// `RestartRequested` event is raised when there is a need to restart WebView2 process
  /// in order to apply certain beneifical updates.
  /// Those updates can be the following cases:
  /// 1. New runtime version available.
  /// 2. Older runtime version available (downgradge due to severe bug).
  /// 3. Features being enabled/disabled that causing WebView2 reliability or performance 
  ///    to drop dramatically. 
  /// 
  /// `RestartRequested` is raised base on priority level. `RestartPriorityLevel` is used
  /// to determine the urgences on rasing such event.
  /// `COREWEBVIEW2_RESTART_PRIORITY_LEVEL_BEST_EFFORT`:
  ///   - Event rasied only on new runtime version available.
  ///   - Developer can notify user to restart on normal cadence.
  /// `COREWEBVIEW2_RESTART_PRIORITY_LEVEL_CRITICAL`: 

>   ///   - Event raised for downgraded runtime version and feature enabled/disabled.

Reviewer: For this and line 138 - we don't want to document that these enum values correspond directly to best effort = upgrade, and critical = downgrade or configuration change. We want to be able to change that in the future

```c#
namespace Microsoft.Web.WebView2.Core
{
    enum CoreWebView2RestartRequestedPriority
    {
        BestEffort = 0,

>         Critical = 1,

Reviewer: The term Critical has a special meaning at least for [Windows Updates](https://learn.microsoft.com/en-us/troubleshoot/windows-client/installing-updates-features-roles/standard-terminology-software-updates) which doesn't apply and would be confusing.

And BestEffort is less clear to me where it would lie in terms of normal priority terms (low, normal, high).

Can we rename (BestEffort, Critical) to (Normal, High) or something along those lines?

Reviewer 2: I agree with Normal and High

Author: Updated.

```c#
namespace Microsoft.Web.WebView2.Core
{
    enum CoreWebView2RestartPriorityLevel
    {
        BestEffort = 0,
        Critical = 1,
    };

    runtimeclass CoreWebView2Environment
    {
        // ...
        event Windows.Foundation.TypedEventHandler<CoreWebView2Environment, Object> RestartRequested;

        // Set restart priority level

>         CoreWebView2RestartPriorityLevel RestartPriorityLevel { get; set; };

Reviewer: I thought this was a get only property on the RestartRequested event args?

Author: Synced offline. Will make the change

>         CoreWebView2RestartPriorityLevel RestartPriorityLevel { get; set; };

Reviewer: Can we call the enum 'CoreWebView2RestartRequestedPriority' and name the property just 'Priority'? We don't need to say 'Restart' in the property name if its on the RestartRequestedEventArgs since then Restart is in the class name already. And 'Level' is sort of unnecessary
